packages/shared-ini-file-loader/src/slurpFile.ts

```diff
@@ -0,0 +1,62 @@
+// ToDo: Change to "fs/promises" when supporting nodejs>=14
+import { promises as fsPromises } from "fs";
+
+const { readFile, stat } = fsPromises;
+
+type callbacks = { resolve: Function; reject: Function };
+type fileStatus = {
+  contents: string;
+  isReading: boolean;
+  lastModified: Date;
+  requestQueue: callbacks[];
+};
+
+const fileReadHash: { [key: string]: fileStatus } = {};
+
+const callReadFile = (path: string, cbs: callbacks) => {
+  fileReadHash[path].requestQueue.push(cbs);
+  readFile(path, "utf8")
+    .then((data) => {
+      // File read successful
+      fileReadHash[path].contents = data;
+      fileReadHash[path].requestQueue.forEach(({ resolve }) => resolve(data));
+    })
+    .catch((err) => {
+      fileReadHash[path].requestQueue.forEach(({ reject }) => reject(err));
+    })
+    .finally(() => {
+      fileReadHash[path].isReading = false;
+      fileReadHash[path].requestQueue = [];
+      // Update last modified.
+      stat(path).then(({ mtime }) => {
+        fileReadHash[path].lastModified = mtime;
+      });
+    });
+};
+
+export const slurpFile = (path: string) =>
+  new Promise<string>((resolve, reject) => {
+    if (!fileReadHash[path]) {
+      // File not read yet, set file isReading to true and read file.
+      fileReadHash[path] = { isReading: true, lastModified: new Date(0), contents: "", requestQueue: [] };
+      callReadFile(path, { resolve, reject });
+    } else if (fileReadHash[path].isReading) {
+      // File currently being read. Add callbacks to the request queue.
+      fileReadHash[path].requestQueue.push({ resolve, reject });
+    } else {
+      // File read was attempted in the past.
+      if (fileReadHash[path].lastModified.getTime() === 0) {
+        // Previous read attempt failed. Attempt read file again.
+        callReadFile(path, { resolve, reject });
+      } else {
+        // Read file only if it's modified.
+        stat(path).then(({ mtime }) => {
+          if (mtime === fileReadHash[path].lastModified) {
+            resolve(fileReadHash[path].contents);
+          } else {
+            callReadFile(path, { resolve, reject });
```

Comment:
Is this a requirement / supported feature?
The file should be read again if it changed?
So in theory one could change the config or credential file in runtime, and for the next client instantiation new configs/credentials would be used?

If this is not a requirement, the additional stat and readFile call could be omitted, right?

Reply:
Not for the cases where SDK is used in Lambda - or an application which runs for a short time.
It would be required if application is long running - like a server application in a container for example.